pkg/stanza/fileconsumer/internal/tracker/tracker.go

```diff
@@ -125,6 +143,34 @@ func (t *fileTracker) TotalReaders() int {
 	return total
 }
 
+func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+	// We make use of a ring buffer, where each set of files is stored under a specific index.
+	// Instead of discarding knownFiles[2], write it to the next index and eventually roll over.
+	// Separate storage keys knownFilesArchive0, knownFilesArchive1, ..., knownFilesArchiveN, roll over back to knownFilesArchive0
+
+	// Archiving:  ┌─────────────────────on-disk archive─────────────────────────┐
+	//             |    ┌───┐     ┌───┐                     ┌──────────────────┐ |
+	// index       | ▶  │ 0 │  ▶  │ 1 │  ▶      ...       ▶ │ polls_to_archive │ |
+	//             | ▲  └───┘     └───┘                     └──────────────────┘ |
+	//             | ▲    ▲                                                ▼     |
+	//             | ▲    │ Roll over overriting older offsets, if any     ◀     |
+	//             └──────│──────────────────────────────────────────────────────┘
+	//                    │
+	//                    │
+	//                    │
+	//                   start
+	//                   index
+
+	if t.pollsToArchive <= 0 || t.persister == nil {
+		return
+	}
```

Comment:
Do we have to worry about these cases? Aren't we checking them and instantiating a `NoStateTracker` when these would ever be relevant?

Reply:
`NoStateTracker` will be created when `WithNoTracking` option is selected. 
We can create a normal tracker without archiving disabled, by setting `pollsToArchive: 0`.
A tracker with archiving enabled will have `pollsToArchive` > `0`. 

The check `t.persister == nil` is just a precautionary check.

Reply:
Alright. Maybe in another PR we can clean this up. IMO we have starting to have too many special cases to handle and the easiest solution to that is move to a pattern where the nils pieces are nops instead. Then the implementation doesn't need to worry about details so much.